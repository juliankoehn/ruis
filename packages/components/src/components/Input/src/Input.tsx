import React, { useState, useRef } from 'react';
import { styles } from './styles';
import { InputProps } from './types';

function useCombinedRefs<T>(...refs: any) {
  const targetRef = React.useRef<T>();

  React.useEffect(() => {
    refs.forEach(
      (ref: React.RefCallback<T> | React.MutableRefObject<T> | null) => {
        if (!ref) return null;

        if (typeof ref === 'function') {
          ref(targetRef.current as T);
        } else {
          ref.current = targetRef.current as T;
        }
      },
    );
  }, [refs]);

  return targetRef as React.RefObject<T>;
}

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref) => {
    const {
      disabled,
      readOnly,
      required,
      invalid,
      onMouseDown,
      onFocus,
      onBlur,
      elemAfterInput,
      elemBeforeInput,

      ...rest
    } = props;
    const [hovered, setHovered] = useState(false);
    const [isFocused, setFocused] = useState(false);

    const inputRef = useRef<HTMLInputElement>();
    const combinedRef = useCombinedRefs<HTMLInputElement>(ref, inputRef);

    const onMouseDownHandler = (event: React.MouseEvent<HTMLElement>) => {
      /** Running e.preventDefault() on the INPUT prevents double click behaviour */
      // Sadly we needed this cast as the target type is being correctly set
      const target: HTMLInputElement = event.target as HTMLInputElement;
      if (target.tagName !== 'INPUT') {
        event.preventDefault();
      }
      if (
        inputRef?.current &&
        !disabled &&
        document.activeElement !== inputRef.current
      ) {
        inputRef.current.focus();
      }
      if (onMouseDown) {
        onMouseDown(event);
      }
    };

    const onMouseEnter = () => {
      if (!disabled) {
        setHovered(true);
      }
    };

    const onMouseLeave = () => {
      if (!disabled) {
        setHovered(false);
      }
    };

    const handleOnFocus = (event: React.FocusEvent<HTMLInputElement>) => {
      setFocused(true);
      if (onFocus) {
        onFocus(event);
      }
    };
    const handleOnBlur = (event: React.FocusEvent<HTMLInputElement>) => {
      setFocused(false);
      if (onBlur) {
        onBlur(event);
      }
    };

    return (
      <div
        css={styles.container({
          disabled: !!disabled,
          hovered: hovered,
          focused: isFocused,
          invalid: invalid,
        })}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onMouseDown={onMouseDownHandler}
      >
        {elemBeforeInput && <span css={styles.before}>{elemBeforeInput}</span>}
        <input
          ref={combinedRef}
          css={styles.input}
          {...rest}
          disabled={disabled ? !!true : undefined}
          readOnly={readOnly}
          required={required}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
        />
        {elemAfterInput && <span css={styles.after}>{elemAfterInput}</span>}
      </div>
    );
  },
);

Input.displayName = 'Input';
