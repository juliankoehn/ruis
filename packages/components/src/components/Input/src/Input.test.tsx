import React from 'react';
import { render } from '@testing-library/react';
import { Input } from './Input';

describe('Input', () => {
  it('should render', () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });

  it('should forward ref', () => {
    const inputRef = React.createRef<HTMLInputElement>();

    render(<Input ref={inputRef} />);

    expect(inputRef.current).not.toBeUndefined();
  });

  it('should be disabled', () => {
    const inputRef = React.createRef<HTMLInputElement>();
    render(<Input ref={inputRef} disabled />);

    expect(inputRef.current).not.toBeUndefined();
    expect(inputRef.current?.hasAttribute('disabled')).toBeTruthy();
  });

  it('should be readOnly', () => {
    const inputRef = React.createRef<HTMLInputElement>();
    render(<Input ref={inputRef} readOnly />);

    expect(inputRef.current).not.toBeUndefined();
    expect(inputRef.current?.hasAttribute('readOnly')).toBeTruthy();
  });

  it('should be required', () => {
    const inputRef = React.createRef<HTMLInputElement>();
    render(<Input ref={inputRef} required />);

    expect(inputRef.current).not.toBeUndefined();
    expect(inputRef.current?.hasAttribute('required')).toBeTruthy();
  });

  it('should be focused', () => {
    const focusSpy = jest.fn();
    const blurSpy = jest.fn();

    const inputRef = React.createRef<HTMLInputElement>();
    const { container } = render(
      <Input ref={inputRef} onFocus={focusSpy} onBlur={blurSpy} />,
    );
    expect(focusSpy).toHaveBeenCalledTimes(0);
    expect(blurSpy).toHaveBeenCalledTimes(0);
    container.querySelector('input')?.focus();
    expect(focusSpy).toHaveBeenCalledTimes(1);
    container.querySelector('input')?.blur();
    expect(blurSpy).toHaveBeenCalledTimes(1);
  });
});
