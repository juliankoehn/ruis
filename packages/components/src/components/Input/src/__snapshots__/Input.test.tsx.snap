// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render 1`] = `
<div>
  <div
    class="css-50aab3"
  >
    <input
      class="css-1tbfif6-input"
    />
  </div>
</div>
`;
